lib/srv/app/server.go

```diff
@@ -0,0 +1,510 @@
+	// If a cached forwarder exists, return it right away.
+	session, err := s.cache.get(identity.RouteToApp.SessionID)
+	if err == nil {
+		return session, nil
+	}
+
+	// Create a new session with a recorder and forwarder in it.
+	session, err = s.newSession(ctx, identity, app)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Put the session in the cache so the next request can use it for 5 minutes
+	// or the time until the certificate expires, whichever comes first.
+	ttl := utils.MinTTL(identity.Expires.Sub(s.c.Clock.Now()), 5*time.Minute)
+	err = s.cache.set(identity.RouteToApp.SessionID, session, ttl)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return session, nil
+}
+
+// getApp returns an application matching the public address. If multiple
+// matching applications exist, the first one is returned. Random selection
+// (or round robin) does not need to occur here because they will all point
+// to the same target address. Random selection (or round robin) occurs at the
+// web proxy to load balance requests to the application service.
+func (s *Server) getApp(ctx context.Context, publicAddr string) (*services.App, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, a := range s.server.GetApps() {
+		if publicAddr == a.PublicAddr {
+			return a, nil
+		}
+	}
+	return nil, trace.NotFound("no application at %v found", publicAddr)
+}
+
+// newHTTPServer creates an *http.Server that can authorize and forward
+// requests to a target application.
+func (s *Server) newHTTPServer() *http.Server {
+	// Reuse the auth.Middleware to authorize requests but only accept
+	// certificates that were specifically generated for applications.
+	authMiddleware := &auth.Middleware{
+		AccessPoint:   s.c.AccessPoint,
+		AcceptedUsage: []string{teleport.UsageAppsOnly},
+	}
+	authMiddleware.Wrap(s)
+
+	return &http.Server{
+		Handler:           authMiddleware,
+		ReadHeaderTimeout: defaults.DefaultDialTimeout,
+	}
+}
+
+// getProxyPort tries to figure out the address the proxy is running at.
+func (s *Server) getProxyPort() string {
+	servers, err := s.c.AccessPoint.GetProxies()
+	if err != nil {
+		return strconv.Itoa(defaults.HTTPListenPort)
+	}
+	if len(servers) == 0 {
+		return strconv.Itoa(defaults.HTTPListenPort)
+	}
+	_, port, err := net.SplitHostPort(servers[0].GetPublicAddr())
+	if err != nil {
+		return strconv.Itoa(defaults.HTTPListenPort)
+	}
+	return port
+}
+
+// getConfigForClient returns the list of CAs that could have signed the
+// client's certificate.
+func (s *Server) getConfigForClient(info *tls.ClientHelloInfo) (*tls.Config, error) {
+	var clusterName string
+	var err error
+
+	// Try and extract the name of the cluster that signed the client's certificate.
+	if info.ServerName != "" {
+		clusterName, err = auth.DecodeClusterName(info.ServerName)
+		if err != nil {
+			if !trace.IsNotFound(err) {
+				s.log.Debugf("Ignoring unsupported cluster name %q.", info.ServerName)
+				clusterName = ""
```

Comment:
Wouldn't clusterName be an empty string anyway in this block, since the error is returned above?